plugins/auth-backend/src/providers/cloudflare-access/provider.ts

```diff
@@ -0,0 +1,255 @@
+   */
+  iss: string;
+  /**
+   * `custom` contains SAML attributes in the Application Token specified by an
+   * administrator in the identity provider configuration.
+   */
+  custom: string;
+};
+
+/** @public */
+export type CloudflareAccessResult = {
+  fullProfile: PassportProfile;
+  expiresInSeconds?: number;
+};
+
+export type CloudflareAccessProviderInfo = {
+  /**
+   * Expiry of the access token in seconds.
+   */
+  expiresInSeconds?: number;
+};
+
+export type CloudflareAccessResponse =
+  AuthResponse<CloudflareAccessProviderInfo>;
+
+export class CloudflareAccessAuthProvider implements AuthProviderRouteHandlers {
+  private readonly teamName: string;
+  private readonly resolverContext: AuthResolverContext;
+  private readonly authHandler: AuthHandler<CloudflareAccessResult>;
+  private readonly signInResolver: SignInResolver<CloudflareAccessResult>;
+  private readonly jwtKeySet: any;
+
+  constructor(options: Options) {
+    this.teamName = options.teamName;
+    this.authHandler = options.authHandler;
+    this.signInResolver = options.signInResolver;
+    this.resolverContext = options.resolverContext;
+    this.jwtKeySet = createRemoteJWKSet(
+      new URL(
+        `https://${this.teamName}.cloudflareaccess.com/cdn-cgi/access/certs`,
+      ),
+    );
+  }
+
+  frameHandler(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    // ProxiedSignInPage calls `/refresh` implicitly each time the backstage
+    // app is refreshed on the browser.
+    // User authentication is then checked here.
+    const result = await this.getResult(req);
+    const response = await this.handleResult(result);
+    res.json(response);
+  }
+
+  start(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  private async getResult(
+    req: express.Request,
+  ): Promise<CloudflareAccessResult> {
+    // JWTs generated by Access are available in a request header as
+    // Cf-Access-Jwt-Assertion and as cookies as CF_Authorization.
+    let jwt = req.header(CF_JWT_HEADER);
+    if (!jwt) {
+      jwt = req.cookies.CF_Authorization;
+    }
+    if (!jwt) {
+      // Only throw if both are not provided by Cloudflare Access since either
+      // can be used.
+      throw new AuthenticationError(
+        `Missing ${CF_JWT_HEADER} from Cloudflare Access`,
+      );
+    }
+
+    // Cloudflare signs the JWT using the RSA Signature with SHA-256 (RS256).
+    // RS256 follows an asymmetric algorithm; a private key signs the JWTs and
+    // a separate public key verifies the signature.
+    const verifyResult = await jwtVerify(jwt, this.jwtKeySet, {
+      // Cloudflare signs the JWT using the RSA Signature with SHA-256 (RS256).
+      algorithms: ['RS256'],
+      issuer: `https://${this.teamName}.cloudflareaccess.com`,
+    });
+    const claims = verifyResult.payload as CloudflareAccessClaims;
+
+    const fullProfile: PassportProfile = {
```

Comment:
There's really no need to provide this full profile tbh, I think you could go with including the claims and original JWT in the result instead. That way we forward all the available information, but it's still quite convenient to use.

Reply:
ok, just gonna return claims directly.